### PR TITLE
[ML] Add Japanese support for the ML / text embedding #95380

### PR DESCRIPTION
**Pull request by daixque**
This PR is now ready to review.

# Overview

This PR will add Japanese text support for text embedding in ML. 

# Current Implementation

At this moment, even if a model which is trained with Japanese Tokenizer is used to text embedding (for vector search), Elasticsearch uses BertTokenizer to tokenize Japanese incoming text with the punctuation option enabled.

For instance, if below text is given:

```
日本語のテスト
```

It will split to: 日/本/語/の/テ/ス/ト, it looks eventually uni-gram tokenizer.

But it should be: 日本/語/の/テスト

Otherwise each token won't appear in the vocabulary which is paired with the trained model.

## API

For Japanese tokenizer, I propose `bert_ja` tokenization configuration on trained model API to identify the model is trained for Japanese language.

```json
      "inference_config": {
        "text_embedding": {
          "vocabulary": { "index": ".ml-inference-native-000001" },
          "tokenization": {
            "bert_ja": {
              "do_lower_case": false,
              "with_special_tokens": true,
              "max_sequence_length": 512,
              "truncate": "first",
              "span": -1
            }
          }
        }
      }
```

# What this PR will change

This PR brings JapaneseTokenizer from`org.apache.lucene.analysis.kuromoji` module to tokenize Japanese text by using the morphological analysis technique.

I'm focusing on BERT model with Japanese. [BertJapaneseTokenizer from BERT project](https://huggingface.co/docs/transformers/model_doc/bert-japanese) (which is written in python) uses MeCab as the pre-tokenizer. 

Japanese (and Chinese/Korean)  morphological analyzer requires a word dictionary to analyze. So lucene has “mecab-ipadic-2.7.0-20070801” dictionary for Japanese. See links below:

- https://github.com/elastic/elasticsearch/blob/main/plugins/analysis-kuromoji/licenses/lucene-NOTICE.txt#L117
- https://github.com/apache/lucene/blob/bed07c6b0242646306470df9265469f6a811b5b2/gradle/generation/kuromoji.gradle#L50
This dictionary is created for MeCab which is another Japanese morphological analyzer implementation than Kuromoji, but if same dictionary is used, so result should be similar if both sides use it.

But, [BertJapaneseTokenizer](https://github.com/cl-tohoku/bert-japanese#tokenization) uses an another dictionary called “Unidic 2.1.2" which is also an outcome of MeCab project. So I believe we can expect tokenization will similar, or at least better than current uni-gram-like implementation.

# Limitation

As it's mentioned above, different Japanese dictionary is used between Elasticsearch side and Pytorch model side. So embedded vector representation would be different in some cases. But I tested with @davidkyle and confirmed differences are very small, so it's acceptable.

# Request for review

The points below are strongly needed to review.

- This PR adds a dependency to `org.apache.lucene.analysis.kuromoji`. It contains Japanese dictionary, so ES binary footprint will be bigger.
- This PR proposes to add new `bert_ja` tokenization configuration on trained model API

# Uploading Japanese model

If this PR is merged, then Eland should be updated as well to be able to set proper configuration. Eland change [here](https://github.com/elastic/eland/pull/534).

# Relates

- https://github.com/elastic/elasticsearch/issues/95380
- https://github.com/elastic/eland/pull/534

### CODE REVIEW

**Comment by daixque** [comment]
@davidkyle This PR is now ready to review. Please take a look.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/BertJapaneseTokenizationUpdate.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.io.IOException;
+import java.util.Optional;
+
+public class BertJapaneseTokenizationUpdate extends AbstractTokenizationUpdate {
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/BertJapaneseTokenizationUpdate.java:20]: I think there is potential to reduce some of the code duplication by reusing the `BertTokenizationUpdate`. But that is not necessary for this PR.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BertJapaneseTokenizerTests.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        try (
+            BertTokenizer tokenizer = BertJapaneseTokenizer.builder(
+                TEST_CASED_VOCAB,
+                new BertTokenization(null, false, null, Tokenization.Truncate.NONE, -1)
+            ).build()
+        ) {
+
+            String msg = "日本語で、ElasticsearchのBertJapaneseTokenizerを使うテスト。";
+            // List<TokenizationResult.Tokens> tokens = tokenizer.tokenize(msg, Tokenization.Truncate.NONE, 1, 0);
+            // System.out.println(List.of(tokens));
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BertJapaneseTokenizerTests.java:58]: nit: please delete the commented out code.

**Review by davidkyle — COMMENTED** [review]
This is great thanks for making the PR @daixque 

I would like to do some testing now and compare the results of evaluating Japanese BERT with the huggingface implementation.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BertJapaneseTokenizerTests.java (hunk 2 above)
Comment (daixque) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BertJapaneseTokenizerTests.java]: Deleted

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/BertJapaneseTokenizationUpdate.java (hunk 1 above)
Comment (daixque) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/BertJapaneseTokenizationUpdate.java:20]: If you mean `BertJapaneseTokenizationUpdate` should be inherited from `BertTokenizationUpdate`, then I have a concern to do that. Assume that there is a instance of BertJapaneseTokenizationUpdate called `ja`, then that `ja` object is also  an instance of BertTokenizationUpdate. So if we write a code for TokenizationUpdate like below in the future, then I'm afraid it potentially become problematic.

https://github.com/elastic/elasticsearch/blob/8e7d4b075092869dbae8e9a9511c13adbb62c30f/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/NlpTokenizer.java#L390-L398

https://github.com/daixque/elasticsearch/blob/5890fca9214c2c823a5c513fb0f42c602ab8ce7a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/NlpTokenizer.java#L391-L402

**Comment by davidkyle** [comment]
@elasticsearchmachine update branch

**Comment by davidkyle** [comment]
@elasticmachine update branch

**Comment by Matzr-4** [comment]
Generate Facebook Authenticator code 

Sent from Yahoo Mail for iPhone


On Sunday, June 11, 2023, 9:19 AM, David Kyle ***@***.***> wrote:




@elasticmachine update branch

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by davidkyle** [comment]
Thanks for your contribution @daixque it's great to have Japanese language support

We would like to make this feature **Technical Preview** due to the stated limitation on the use of different Japanese dictionary to the one used in training.

**Pull request merged**